### ISSUE TITLE
SCAPP failing during classify stage

### ISSUE THREAD

**Issue by rhysnewell** [Shamir-Lab/SCAPP#17]
Hi!

Hope you are doing well, just ran into a bug when using the conda version of scapp.
installed via `conda crate -n scapp -c bioconda scapp`

I noticed I was not getting any output when running scapp, instead there was an error being thrown in one of the log files. The error can be reproduced by using the test scripts provided on your github.

### Bug:
When running the `./run_test.sh` script no output is created. Error in `logs/plasclass_std.log`

### Output:
```
Loading classifier 1000
/home/n10853499/.local/lib/python3.7/site-packages/sklearn/utils/deprecation.py:143: FutureWarning: The sklearn.linear_model.logistic module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.linear_model. Anything that cannot be imported from sklearn.linear_model is now part of the private API.
  warnings.warn(message, FutureWarning)
/home/n10853499/.local/lib/python3.7/site-packages/sklearn/base.py:334: UserWarning: Trying to unpickle estimator LogisticRegression from version 0.19.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/home/n10853499/.local/lib/python3.7/site-packages/sklearn/utils/deprecation.py:143: FutureWarning: The sklearn.preprocessing.data module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.preprocessing. Anything that cannot be imported from sklearn.preprocessing is now part of the private API.
  warnings.warn(message, FutureWarning)
/home/n10853499/.local/lib/python3.7/site-packages/sklearn/base.py:334: UserWarning: Trying to unpickle estimator StandardScaler from version 0.19.1 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
Loading classifier 10000
Loading classifier 100000
Loading classifier 500000
712 sequences to classify. Classifying in batches of 100k
Starting new batch
Getting kmer frequencies for partition length 1000
Traceback (most recent call last):
  File "/home/n10853499/.conda/envs/scapp/bin/scapp", line 10, in <module>
    sys.exit(main())
  File "/home/n10853499/.conda/envs/scapp/lib/python3.7/site-packages/scapp/scapp.py", line 285, in main
    classify_fastg.classify(fastg, plasclass_file, num_procs)
  File "/home/n10853499/.conda/envs/scapp/lib/python3.7/site-packages/scapp/classify_fastg.py", line 56, in classify
    probs = c.classify(seqs)
  File "/home/n10853499/.conda/envs/scapp/lib/python3.7/site-packages/plasclass/plasclass.py", line 62, in classify
    standardized_freqs = self._standardize(kmer_freqs_mat, scale)
  File "/home/n10853499/.conda/envs/scapp/lib/python3.7/site-packages/plasclass/plasclass.py", line 105, in _standardize
    return self.classifiers[scale]['scaler'].transform(freqs)
  File "/home/n10853499/.local/lib/python3.7/site-packages/sklearn/preprocessing/_data.py", line 794, in transform
    force_all_finite='allow-nan')
  File "/home/n10853499/.local/lib/python3.7/site-packages/sklearn/base.py", line 436, in _validate_data
    self._check_n_features(X, reset=reset)
  File "/home/n10853499/.local/lib/python3.7/site-packages/sklearn/base.py", line 373, in _check_n_features
    "The reset parameter is False but there is no "
RuntimeError: The reset parameter is False but there is no n_features_in_ attribute. Is this estimator fitted?
```

Any help would be appreciated!

Cheers,
Rhys

**Comment by dpellow** [comment]
I haven't yet been able to replicate this on my system but will keep trying. In the meantime can you let me know the versions of the numpy, scipy, and scikit-learn packages in your environment? You can see them with `conda list`.
Also - what OS and python version are you using?

**Comment by dpellow** [comment]
@rhysnewell  - I wasn't able to replicate this error. You installed SCAPP from bioconda correct?

In addition to the above can you let me know more details about the system - RAM, processors?

From what I've found for similar errors online, my guess it has something to do with the version of scikit-learn, but the recipe should pin it to version 0.21.3 - do you have something different?

**Comment by rhysnewell** [comment]
@dpellow I've been trying to run it on a linux system with 128 cpus and 1 TB RAM, here is the OS release:

```
NAME="SLES"
VERSION="12-SP5"
VERSION_ID="12.5"
PRETTY_NAME="SUSE Linux Enterprise Server 12 SP5"
ID="sles"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:12:sp5"
```

 Here is the environment I have:
```
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                      1_llvm    conda-forge
blas                      2.17                   openblas    conda-forge
blast                     2.10.1          pl526he19e7b1_2    bioconda
bwa                       0.7.17               hed695b0_7    bioconda
bzip2                     1.0.8                h516909a_3    conda-forge
c-ares                    1.11.0               h470a237_1    bioconda
ca-certificates           2020.6.20            hecda079_0    conda-forge
certifi                   2020.6.20        py37he5f6b98_2    conda-forge
curl                      7.71.1               he644dc0_8    conda-forge
decorator                 4.4.2                      py_0    conda-forge
entrez-direct             13.9            pl526h375a9b1_0    bioconda
expat                     2.2.9                he1b5a44_2    conda-forge
joblib                    0.14.1             pyh9f0ad1d_0    conda-forge
krb5                      1.17.1               hfafb76e_3    conda-forge
ld_impl_linux-64          2.35                 h769bd43_9    conda-forge
libblas                   3.8.0               17_openblas    conda-forge
libcblas                  3.8.0               17_openblas    conda-forge
libcurl                   7.71.1               hcdd3856_8    conda-forge
libdeflate                1.0                  h14c3975_1    bioconda
libedit                   3.1.20191231         h46ee950_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.2.1             he1b5a44_1007    conda-forge
libgcc-ng                 9.3.0               h5dbcf3e_17    conda-forge
libgfortran-ng            7.5.0               hae1eefd_17    conda-forge
libgfortran4              7.5.0               hae1eefd_17    conda-forge
liblapack                 3.8.0               17_openblas    conda-forge
liblapacke                3.8.0               17_openblas    conda-forge
libnghttp2                1.41.0               hab1572f_1    conda-forge
libopenblas               0.3.10          pthreads_hb3c22a3_5    conda-forge
libssh2                   1.9.0                hab1572f_5    conda-forge
libstdcxx-ng              9.3.0               h2ae2ef3_17    conda-forge
llvm-openmp               11.0.0               hfc4b9b4_1    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
networkx                  2.4                        py_1    conda-forge
numpy                     1.17.4           py37hd5be1e1_0
numpy-base                1.17.4           py37h2f8d375_0
openssl                   1.1.1h               h516909a_0    conda-forge
pcre                      8.44                 he1b5a44_0    conda-forge
perl                      5.26.2            h36c2ea0_1008    conda-forge
perl-app-cpanminus        1.7044                  pl526_1    bioconda
perl-archive-tar          2.32                    pl526_0    bioconda
perl-base                 2.23                    pl526_1    bioconda
perl-business-isbn        3.004                   pl526_0    bioconda
perl-business-isbn-data   20140910.003            pl526_0    bioconda
perl-carp                 1.38                    pl526_3    bioconda
perl-common-sense         3.74                    pl526_2    bioconda
perl-compress-raw-bzip2   2.087           pl526he1b5a44_0    bioconda
perl-compress-raw-zlib    2.087           pl526hc9558a2_0    bioconda
perl-constant             1.33                    pl526_1    bioconda
perl-data-dumper          2.173                   pl526_0    bioconda
perl-digest-hmac          1.03                    pl526_3    bioconda
perl-digest-md5           2.55                    pl526_0    bioconda
perl-encode               2.88                    pl526_1    bioconda
perl-encode-locale        1.05                    pl526_6    bioconda
perl-exporter             5.72                    pl526_1    bioconda
perl-exporter-tiny        1.002001                pl526_0    bioconda
perl-extutils-makemaker   7.36                    pl526_1    bioconda
perl-file-listing         6.04                    pl526_1    bioconda
perl-file-path            2.16                    pl526_0    bioconda
perl-file-temp            0.2304                  pl526_2    bioconda
perl-html-parser          3.72            pl526h6bb024c_5    bioconda
perl-html-tagset          3.20                    pl526_3    bioconda
perl-html-tree            5.07                    pl526_1    bioconda
perl-http-cookies         6.04                    pl526_0    bioconda
perl-http-daemon          6.01                    pl526_1    bioconda
perl-http-date            6.02                    pl526_3    bioconda
perl-http-message         6.18                    pl526_0    bioconda
perl-http-negotiate       6.01                    pl526_3    bioconda
perl-io-compress          2.087           pl526he1b5a44_0    bioconda
perl-io-html              1.001                   pl526_2    bioconda
perl-io-socket-ssl        2.066                   pl526_0    bioconda
perl-io-zlib              1.10                    pl526_2    bioconda
perl-json                 4.02                    pl526_0    bioconda
perl-json-xs              2.34            pl526h6bb024c_3    bioconda
perl-libwww-perl          6.39                    pl526_0    bioconda
perl-list-moreutils       0.428                   pl526_1    bioconda
perl-list-moreutils-xs    0.428                   pl526_0    bioconda
perl-lwp-mediatypes       6.04                    pl526_0    bioconda
perl-lwp-protocol-https   6.07                    pl526_4    bioconda
perl-mime-base64          3.15                    pl526_1    bioconda
perl-mozilla-ca           20180117                pl526_1    bioconda
perl-net-http             6.19                    pl526_0    bioconda
perl-net-ssleay           1.88            pl526h90d6eec_0    bioconda
perl-ntlm                 1.09                    pl526_4    bioconda
perl-parent               0.236                   pl526_1    bioconda
perl-pathtools            3.75            pl526h14c3975_1    bioconda
perl-scalar-list-utils    1.52            pl526h516909a_0    bioconda
perl-socket               2.027                   pl526_1    bioconda
perl-storable             3.15            pl526h14c3975_0    bioconda
perl-test-requiresinternet 0.05                    pl526_0    bioconda
perl-time-local           1.28                    pl526_1    bioconda
perl-try-tiny             0.30                    pl526_1    bioconda
perl-types-serialiser     1.0                     pl526_2    bioconda
perl-uri                  1.76                    pl526_0    bioconda
perl-www-robotrules       6.02                    pl526_3    bioconda
perl-xml-namespacesupport 1.12                    pl526_0    bioconda
perl-xml-parser           2.44            pl526h4e0c4b3_7    bioconda
perl-xml-sax              1.02                    pl526_0    bioconda
perl-xml-sax-base         1.09                    pl526_0    bioconda
perl-xml-sax-expat        0.51                    pl526_3    bioconda
perl-xml-simple           2.25                    pl526_1    bioconda
perl-xsloader             0.24                    pl526_0    bioconda
pip                       20.2.4                     py_0    conda-forge
plasclass                 0.1.0                      py_0    bioconda
pysam                     0.15.3           py37hda2845c_1    bioconda
python                    3.7.8           h6f2ec95_1_cpython    conda-forge
python_abi                3.7                     1_cp37m    conda-forge
readline                  8.0                  h46ee950_1    conda-forge
samtools                  1.9                 h8571acd_11    bioconda
scapp                     0.1.4                      py_0    bioconda
scikit-learn              0.21.3           py37hcdab131_0    conda-forge
scipy                     1.3.2            py37h921218d_0    conda-forge
setuptools                49.6.0           py37he5f6b98_2    conda-forge
sqlite                    3.32.3               hcee41ef_1    conda-forge
tk                        8.6.10               hed695b0_1    conda-forge
wheel                     0.35.1             pyh9f0ad1d_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
```

**Comment by dpellow** [comment]
OK, there shouldn't be any problem due to the RAM or processors then.

The environment you posted lists scikit-learn 0.21.3 but in the original post, there's a FutureWarning that says it's using version 0.23.2. 
Based on some other similar bug reports (e:g [this one](https://github.com/hyperopt/hyperopt/issues/668#issuecomment-631337171)), it seems like this may have been introduced by something in sklearn v0.23.

Can you tell why it seems to be running 0.23.2 even though it says 0.21.3 is installed? Can you check what version of sklearn is actually used if you open python while in your conda environment and import it?

If it is 0.23.2, what happens if you try: `conda install scikit-learn==0.21.3` in your environment?

**Comment by rhysnewell** [comment]
I've managed to find the cause of the issue. It relates to this problem here https://github.com/conda/conda/issues/9392 
The system conda was not properly appending the my conda environment to my PATH variables, so even though when I ran `which python` it would point to the correct python it would end up importing the wrong python packages.

The solution:
insert `unset CONDA_SHLVL` directly above the `>>> conda initialize >>>` block in my `~/.bashrc` and then reactivate my `SCAPP` environment

Thanks for your help, problem solved!
